### ISSUE TITLE
withMutation should support loading/error/called/data states

### ISSUE THREAD

**Issue by slikts** [apollographql/react-apollo#2952]
Previous (closed) issues: #421 #1842

The `graphql()` (and `withMutation()`) HOCs currently have a surprising limitation in that the mutation results aren't passed by them, which isn't explained by the docs, and there also isn't an obvious reason for it. The natural expectation from reading the docs would be that HOCs and components would have feature parity (despite HOCs having fallen by the wayside to render props and more recently hooks) and would just be alternative APIs for the same functionality, especially since `graphql()` delegates to components under the hood. In other words, passing mutation results in `withMutation()` would follow the principle of least surprise.

The answer also shouldn't be "just use components or hooks"; the reason to pick HOCs currently is that components suffer from "wrapper hell" unless a helper library like react-adopt is used, and that hooks are still immature.

**Comment by cihadturhan** [comment]
I prefer using HOC over components and hooks. With HOC, I can easily separate query/mutate logic from view, especially using with decorators


```js
@graphql(MY_CANDIDATES_AND_PRIMARY, { name: "myCandidatesQuery" })
@graphql(CANDIDATES, { name: "candidatesQuery" })
@graphql(UPDATE_CANDIDATES, { name: "updateCandidatesMutation" })
class CandidatesListScreen extends Component {
   //...
}
```

However, I can't get why there is no variables provided like queries.

**Comment by rwoody** [comment]
I have the same problem when wrapping my components with `graphql`. Isn't this a regression? I thought I remembered mutations triggering `loading` and `error` props...

**Comment by slikts** [comment]
Yes; it is (was) a regression and is fixed by #3008 (not sure in which release the changes are in).

Closing this ticket as a duplicate of #1967

**Comment by rwoody** [comment]
Thanks you!
